src/box/lua/config/applier/sharding.lua

```diff
@@ -0,0 +1,71 @@
+local log = require('internal.config.utils.log')
+local fiber = require('fiber')
+_G.vshard = nil
+
+local function vshard_cfg(configdata)
+    local roles = configdata:get('sharding.roles')
+    if roles == nil then
+        return
+    end
+    if _G.vshard == nil then
+        _G.vshard = require('vshard')
+    end
+    for _, role in pairs(roles) do
+        local cfg = configdata:sharding()
+        cfg.listen = box.cfg.listen
+        cfg.read_only = box.cfg.read_only
+        cfg.replication = box.cfg.replication
+        if role == 'storage' then
+            local replicaset_uuid = configdata._replicaset_uuid
+            assert(replicaset_uuid == box.cfg.replicaset_uuid)
+            local instance_uuid = configdata._instance_uuid
+            assert(instance_uuid == box.cfg.instance_uuid)
+            local this_replicaset_cfg = cfg.sharding[replicaset_uuid]
+            --
+            -- Currently, the replicaset master must set itself as the master in
+            -- its own configuration.
+            -- TODO: remove when vshard introduces auto-discovery of masters.
+            --
+            if not box.info.ro then
+                this_replicaset_cfg.master = nil
+                this_replicaset_cfg.replicas[instance_uuid].master = true
+            end
+            log.info('sharding: apply sharding config')
+            _G.vshard.storage.cfg(cfg, instance_uuid)
+            --
+            -- Currently, replicaset masters may not be aware of all other
+            -- masters, so the rebalancer is disabled.
+            -- TODO: remove when vshard introduces auto-discovery of masters.
+            --
+            if _G.vshard.storage.internal.is_rebalancer_active then
+                log.info('sharding: disable rebalancer')
+                _G.vshard.storage.rebalancer_disable()
+            end
+        end
+        if role == 'router' then
+            _G.vshard.router.cfg(cfg)
+        end
+    end
+end
+
+local function wait_ro_rw(configdata)
+    while true do
+        if box.info.ro then
+            box.ctl.wait_rw()
+        else
+            box.ctl.wait_ro()
+        end
+        vshard_cfg(configdata)
+    end
+end
+
+local function apply(config)
+    local configdata = config._configdata
+    vshard_cfg(configdata)
+    fiber.create(wait_ro_rw, configdata)
+end
```

Comment:
So, we get many fibers that're trying to apply its own config version in some unspecified order. I guess that we should get actual config from a background always-existing fiber. Maybe we also need a check, whether given config version is applied in the given ro/rw state to eliminate races.

Reply:
Should we do this as a separate issue?

Reply:
It looks solved now. Thanks!

I'm not sure, whether we should check the fiber for being in the `dead` state. Let's think on it later.